### DISCUSSION TITLE
tower_http::trace::TraceLayer.on_eos can not be called

### DISCUSSION THREAD

**Discussion by kobe2000** [tokio-rs/axum#2453] [Q&A] [answered]
- [x] I have looked for existing issues (including closed) about this

### Version

* axum 0.7.2
* tower 0.4.13
* tower-http 0.5.0
* tokio 1.35.1
* tracing 0.1.40
* tracing-subscriber 0.3.18

### Platform

Windows x64

### Description

I want to measure the whole time of http server processing (including network IO), so I try to use TraceLayer to log the time like following code:

```rust
let app = Router::new()
        .route("/api", post(api))
        .layer(TraceLayer::new_for_http().on_eos(|trailers: Option<&HeaderMap>, stream_duration: Duration, _span: &Span| {
            println!("stream closed after {:?}", stream_duration)
        }))
```

But I can not see anying on console, please help.

Thank you!

**Comment by davidpdrsn** [comment, marked as answer]
That's not what `on_eos` is intended for. In order to get accurate measurements, including IO, you'll have to use hyper. I'm not sure which APIs hyper has to measure this kinda of stuff but try asking in [hyper's discord channel](https://discord.gg/tokio).
